### PR TITLE
Major Bug Fix 2 for 1.0.0

### LINKED ISSUE #76: Undefined behaviour when cloning game state while a Battlecry is in progress
I have to speak verbosely to explain this one.

You may have asked yourself: Why does MetaStone use EntityReferences everywhere, which are basically just a pointer to a Entity instead of passing plain old Java references to Entities directly? Well the sole reason for this is cloning.

To build any meaningful AI you have to be able to clone the entire game state (for example to try out different moves and check which one performs best). So there is one very important requirement to allow this: You should be able to call gameContext.clone() and get a completely independent copy of the current game state. All changes made to the copy should NOT affect the original game state in any shape or form. 

EntityReferences are a big step towards this goal, but more things have to be to be considered. All objects which are used inside the GameContext need to be either immutable or clonable. And whenever a .clone() method is implemented, the resulting instance has be to be completely independent of the source object. This may sound easy, but in practice it is very easy to forget to clone a list for example. The resulting bugs are incredibly hard to find, as the program compiles just fine and when using an AI which does not clone the GameState, nothing happens. However when an AI like GameStateValue is employed, undefined behaviour may occur.

Recently it came to my attention that there is indeed such a bug left. The problem is the 'environment' field of GameContext. It contains direct references to 'Entity' instances and when cloning the game state, those references still point the old context. Most of the time this is not a problem, as the entries in environment are all temporary and the HashMap should be empty at the end of each move. However, when performing a Battlecry action, this does not hold true. Basically when performing a Batllecry, the player behaviour is prompted for an action in the middle of another action. In this case, the SummonStack will contain a reference to the anew summoned minion.

I am not sure what the best solution for that issue would be. My gut feeling tells me that the 'environment' table itself is evil and should probably go away. However, it helps a lot with implementing some mechanics and I do not know alternatives for all of them (where do we store stuff like event targets, etc).

So we could instead put EntityReferences to environment (and all associated collections like SummonStack, etc). But this won't work right now, as some Entities exists SOLELY inside the environment table, i.e. it is not just a reference but the Entity itself. So we could introduce another List<Entity> in the GameContext, maybe called something like 'pendingEntities' which contains only Entities which do not live anywhere else (on the board, in the players hands or deck, graveyard, etc). Then the environment could contain EntityReferences, which can just be shared across cloned game states. 

If you have any thoughts on this topic, please let me know. Otherwise I might start implementing the solution mentioned above shortly.

### PR DESCRIPTION
**Pull request by webadict**
-Removed Entities from Environment. Replaced with Entity References. (Fixes #76)
-Replaced some Spells. (ReplaceHandSpell and ReplaceDeckSpell)
-Removed some more Spells. (MinMaxDamageSpell and DestroyWeaponSpell)
-Made Gorillabot a Mech and made it so that he didn't accidentally count himself when determining his Battlecry.
-Fixed Wilfred Fizzlebang to accurately modify the Mana Cost of Cards he draws.
-Added a new Spell (AddManaCostModifierSpell)
-Added Logger info to let users know which card file the Card Parser breaks on!
-Changed the description on Cards that were ninja-changed.
-Removed old documentation that did nothing, and added some that actually might be useful.
-Added Event Sources. Changed Events by a lot.
-Removed unnecessary Triggers. Changed Card Triggers to work rather efficiently.
-Changed targeting for some Battlecries to avoid self-Targeting (Looking at you, Tinkmaster...)

EDIT: I've done some testing, and it doesn't appear there's any issues.

### CODE REVIEW

**Comment by demilich1** [comment]
First of all: Thank you, many good stuff in there.

A change that confused me though is the introduction of 'TargetType' enum and its usage in 'hostTargetType'. For example let us look at Axe Flinger:

```
"eventTrigger": {
  			"class": "DamageReceivedTrigger",
			"hostTargetType": "TARGET"
  		},
...
```

For me it is really hard to decipher what that means. I know that you basically merged some boolean  flags into a single enum (which in itself is very sensible), but if I had to create a new card, I would not know what to put there. I think finding more meaningful identifiers would go a long way in this case.

I also got a question regarding the new 'pendingCard' and 'destroyedWeapon' fields. You basically replaced the entries in the 'environment' with those fields, however, this does not solve the problem. When cloning the game state, those fields will also cloned and they contain direct Java references to Entity objects. While they reside no longer inside the environment map, the underlying issue does not seem to be solved. What was your intention behind this change?

**Comment by webadict** [comment]
Fair enough. I did that only as a note to myself. It's really easy to change that to something more recognizable like "AS_TARGET_ONLY", which would be really easy to understand. I'll just do that for all of them, because you are right, it is far too confusing right now.

As for those other changes, I might have forgotten those two minor changes. Let me update all these quickly.

**Comment by webadict** [comment]
Hmm, for the destroyed Weapon, that was added as a safety net for destroying a Weapon. There shouldn't be any direct Environment calls to it, anyhow. It was set because Weapons are added and have their Battlecries go off before processing the destruction of the previous weapon, and in order to keep the Weapon as a findable Entity, it was added to each Player so it could be found when searching for dead Entities (And should act exactly like a Hero's Weapon in every case.) It can be added to the Environment properly later, but for now, no cards actually target destroyed Weapons, so I won't bother.

The pending Card was absolutely needed, since you need to find that particular Card somewhere. I'm not sure how you'll fix it any other way, but the Environment variable is no longer the issue. There needs to exist a carryover Entity, since the Card cannot remain in Hand, cannot be discarded, cannot be set aside... It really can't have anything changed about it.

However, I forgot clone sequences for these, which I've now added.

**Comment by demilich1** [comment]
I did not explain the problem with the 'environment' hashmap very well unfortunately. The basic problem is the following:

When cloning the game state, no instance of Entity (or any subclass for that matter) can be referenced from more than 1 place. It does not matter if it is stored inside a hash table called 'environment' or a field called 'pendingCard'. The problem occurs if it exists in two places: for example the players hand AND in pendingCard. Because when we clone the GameContext in that state, the card now exists twice: We create a copy of the players hand (by cloning each card) AND we create a copy of the 'pendingCard' field. However, if we skip cloning any of those fields, we got a reference to the source GameContext, which we don't want.

So basically we need to make sure that each Entity is only referenced from exactly one place. For all cards it is either the hand, the deck or the graveyard. Minions only live inside their players. Now the field 'pendingCard' would be fine IF the card assigned to 'pendingCard' is NOT in any of the following: the player's hand, deck or graveyard.

But this does not hold true, because when you play a minion with battlecry the card will be assigned to 'pendingCard' and also the card is already in the graveyard. So maybe we could get away with removing the 'pendingCard' field because TargetLogic should find cards inside the graveyard already?

**Comment by webadict** [comment]
I see... Well then, hopefully this is the fix you're looking for? There shouldn't be any duplicates now.
